torchani/data/new.py

```diff
@@ -0,0 +1,458 @@
+
+    # truncate redundant padding - time: 1.3s
+    chunks_batch_species = trunc_pad(list(chunks_batch_species), padding_value=-1)
+    chunks_batch_coordinates = trunc_pad(list(chunks_batch_coordinates))
+
+    for i, c in enumerate(chunks_batch_coordinates):
+        chunks_batch_coordinates[i] = c.reshape(c.shape[0], -1, 3)
+
+    datas = [chunks_batch_species, chunks_batch_coordinates]
+
+    chunks = list(zip(*datas))
+
+    for i, _ in enumerate(chunks):
+        chunks[i] = (chunks[i][0], chunks[i][1])
+
+    properties = {'energies': batch_energies.flatten()}
+
+    # return: [chunk1, chunk2, ...], {"energies", "force", ...} in which chunk1=(species, coordinates)
+    # e.g. chunk1 = [[1807, 21], [1807, 21, 3]], chunk2 = [[193, 50], [193, 50, 3]]
+    # 'energies' = [2000, 1]
+    return chunks, properties
+
+
+def split_to_two_chunks(counts, chunk_threshold):
+
+    counts = counts.cpu()
+    left_mask = torch.triu(torch.ones([counts.shape[0], counts.shape[0]], dtype=torch.uint8))
+    left_mask = torch.transpose(left_mask, dim0=0, dim1=1)
+
+    counts_atoms = counts[:, 0].repeat(counts.shape[0], 1)
+    counts_counts = counts[:, 1].repeat(counts.shape[0], 1)
+
+    counts_atoms_left = torch.where(left_mask, counts_atoms, torch.zeros_like(counts_atoms))
+    counts_atoms_right = torch.where(~left_mask, counts_atoms, torch.zeros_like(counts_atoms))
+    counts_counts_left = torch.where(left_mask, counts_counts, torch.zeros_like(counts_atoms))
+    counts_counts_right = torch.where(~left_mask, counts_counts, torch.zeros_like(counts_atoms))
+
+    # chunk max
+    chunk_max_left, _ = torch.max(counts_atoms_left, dim=-1, keepdim=True)
+    chunk_max_right, _ = torch.max(counts_atoms_right, dim=-1, keepdim=True)
+
+    # chunk size
+    chunk_size_left = torch.sum(counts_counts_left, dim=-1, keepdim=True, dtype=torch.int32)
+    chunk_size_right = torch.sum(counts_counts_right, dim=-1, keepdim=True, dtype=torch.int32)
+
+    # calculate cost
+    min_cost_threshold = torch.tensor([chunk_threshold], dtype=torch.int32)
+    cost = (torch.max(chunk_size_left * chunk_max_left * chunk_max_left, min_cost_threshold)
+            + torch.max(chunk_size_right * chunk_max_right * chunk_max_right, min_cost_threshold))
+
+    # find smallest cost
+    cost_min, cost_min_index = torch.min(cost.squeeze(), dim=-1)
+
+    # find smallest cost chunk_size, if not splitted, it will be [max_chunk_size, 0]
+    final_chunk_size = [chunk_size_left[cost_min_index], chunk_size_right[cost_min_index]]
+    final_chunk_max = [chunk_max_left[cost_min_index], chunk_max_right[cost_min_index]]
+
+    # if not splitted
+    if cost_min_index == (counts.shape[0] - 1):
+        return False, counts, [final_chunk_size[0]], [final_chunk_max[0]], cost_min
+    # if splitted
+    return True, [counts[:cost_min_index + 1], counts[(cost_min_index + 1):]], \
+        final_chunk_size, final_chunk_max, cost_min
+
+
+def split_to_chunks(counts, chunk_threshold=50000):
```

Comment:
This is more or less a duplication with `split_batch`, which is bad, but acceptable. It would be nice if someone could merge these two codes, but it is OK to leave it as is right now.

Reply:
Are you sure?

Reply:
I am saying, we should merge this part with the legacy code in the future. Nothing to do right now.

Reply:
your code will be removed in the future.